src/java.base/share/classes/javax/security/auth/login/LoginContext.java

```diff
@@ -691,21 +714,18 @@ private void invoke(String methodName) throws LoginException {
                     // locate and instantiate the LoginModule
                     //
                     String name = moduleStack[i].entry.getLoginModuleName();
-                    @SuppressWarnings("removal")
-                    ServiceLoader<LoginModule> sc = AccessController.doPrivileged(
-                            (PrivilegedAction<ServiceLoader<LoginModule>>)
-                                    () -> ServiceLoader.load(
-                                        LoginModule.class, contextClassLoader));
-                    for (LoginModule m: sc) {
-                        if (m.getClass().getName().equals(name)) {
-                            moduleStack[i].module = m;
-                            if (debug != null) {
-                                debug.println(name + " loaded as a service");
+                    synchronized(cacheServiceProviders){
+                        Set<Provider<LoginModule>> lmp = cacheServiceProviders.get(contextClassLoader);
+                        for ( Provider<LoginModule> lm: lmp){
+                            if (lm.type().getName().equals(name)){
+                                moduleStack[i].module = lm.get();
```

Comment:
There is a small behavior change here. Originally the login module class is loaded first and then its name is checked. With this change, the name is first checked and then class loaded. This is certainly a performance boost but unfortunately the test at `test/jdk/javax/security/auth/spi/Loader.java` would fail. That test was written to ensure that services provided through a service loader are checked first and then fallback to `Class.forName()`. Either you need to enhance the test to confirm this in a new way, or it would have to be removed.